subsys/bluetooth/services/ots/ots_client.c

```diff
@@ -1156,7 +1196,70 @@ static int oacp_read(struct bt_conn *conn,
 
 	return err;
 }
+static int oacp_write(struct bt_conn *conn,
+		     struct bt_otc_internal_instance_t *inst,
+			 uint8_t *buf, size_t len, off_t offset, uint8_t mode)
+{
+	int err;
+	struct bt_gatt_ots_l2cap *l2cap;
+
+	LOG_DBG("");
+	if (!inst->otc_inst->oacp_handle) {
+		BT_DBG("Handle not set");
+		return -EINVAL;
+	} else if (inst->busy) {
+		return -EBUSY;
+	} else if (cur_inst) {
+		return -EBUSY;
+	}
+
+	/* TODO: How do we ensure that the L2CAP is connected in time for the
+	 * transfer?
+	 */
+
+	err = bt_gatt_ots_l2cap_connect(conn, &l2cap);
+	if (err) {
+		BT_DBG("Could not connect l2cap: %d", err);
+		return err;
+	}
+	l2cap->tx_done = write_obj_tx_done;
+	l2cap->rx_done = rx_done;
+	l2cap->closed  = chan_closed;
+	l2cap->tx.data = buf;
+	l2cap->tx.len = len;
+
+	net_buf_simple_reset(&otc_tx_buf);
+
+	/* OP Code */
+	net_buf_simple_add_u8(&otc_tx_buf, BT_GATT_OTS_OACP_PROC_WRITE);
+
+	/* Offset */
+	net_buf_simple_add_le32(&otc_tx_buf, offset);
+
+	/* Len */
+	net_buf_simple_add_le32(&otc_tx_buf, len);
+	/* Mode, default no truncate*/
+	net_buf_simple_add_u8(&otc_tx_buf, 0);
+
+	/* */
+
+	inst->otc_inst->write_params.offset = 0;
+	inst->otc_inst->write_params.data = otc_tx_buf.data;
+	inst->otc_inst->write_params.length = otc_tx_buf.len;
+	inst->otc_inst->write_params.handle = inst->otc_inst->oacp_handle;
+	inst->otc_inst->write_params.func = write_oacp_cp_write_req_cb;
+
+	err = bt_gatt_write(conn, &inst->otc_inst->write_params);
+
+	if (!err) {
+		inst->busy = true;
+	}
+
+	cur_inst = inst;
```

Comment:
Shouldn't this be move to `if (!err)` too?

Reply:
This logic is straight came from oacp_read().
I think the logic behind this is to let callback function to release instance busy state either ATT queue is full or not.

Reply:
But if there is an error (i.e. `if (!err)` fails), then there isn't any callbacks that are called.

Reply:
This comment is still valid